clippy_utils/src/lib.rs

```diff
@@ -575,25 +582,115 @@ pub fn can_move_expr_to_closure_no_visit(cx: &LateContext<'tcx>, expr: &'tcx Exp
         | ExprKind::LlvmInlineAsm(_) => false,
         // Accessing a field of a local value can only be done if the type isn't
         // partially moved.
-        ExprKind::Field(base_expr, _)
-            if matches!(
-                base_expr.kind,
-                ExprKind::Path(QPath::Resolved(_, Path { res: Res::Local(_), .. }))
-            ) && can_partially_move_ty(cx, cx.typeck_results().expr_ty(base_expr)) =>
-        {
+        ExprKind::Field(
+            &Expr {
+                hir_id,
+                kind:
+                    ExprKind::Path(QPath::Resolved(
+                        _,
+                        Path {
+                            res: Res::Local(local_id),
+                            ..
+                        },
+                    )),
+                ..
+            },
+            _,
+        ) if !ignore_locals.contains(local_id) && can_partially_move_ty(cx, cx.typeck_results().node_type(hir_id)) => {
             // TODO: check if the local has been partially moved. Assume it has for now.
             false
-        }
+        },
         _ => true,
     }
 }
 
-/// Checks if the expression can be moved into a closure as is.
-pub fn can_move_expr_to_closure(cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>) -> bool {
+/// How a local is captured by a closure
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum CaptureKind {
+    Value,
+    Ref(Mutability),
+}
```

Comment:
I'm concerned that this is basically re-implementing `ExprUseVisitor`. Have you tried using that instead? There maybe nuances there that we need to catch like `fake_read`.

Reply:
It implements a subset of that. `fake_read` doesn't need to be considered, that's needed to adjust things for the borrow checker.

I could make use of it. It would mean running two visitors, but I doubt the speed difference would be noticeable even with the extra work `ExprUseVisitor` does.

Reply:
I'll leave it up to you.

Reply:
After getting it to work properly with patterns and nested closures, should have just gone with `ExprUseVisitor`. If `capture_local_usage` is something that would be used somewhere else it's worth it, but I don't see it being used.

Reply:
Are you saying that you might switch to `ExprUseVisitor`?

Reply:
I would say keep it as is. It's already written. It shouldn't need to touched much until captures switch to working on fields which would require changes anyways.